server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -167,22 +164,94 @@ private ContextResource getMailResource()
-				mailResource.setProperty("mail.smtp.user", mailProps.smtpUser);
-				mailResource.setProperty("mail.smtp.port", mailProps.smtpPort);
+				mailResource.setProperty("mail.smtp.host", mailProps.getSmtpHost());
+				mailResource.setProperty("mail.smtp.user", mailProps.getSmtpUser());
+				mailResource.setProperty("mail.smtp.port", mailProps.getSmtpPort());
+
+				if (mailProps.getSmtpFrom() != null &&
+					mailProps.getSmtpPassword() != null &&
+					mailProps.getSmtpStartTlsEnable() != null &&
+					mailProps.getSmtpSocketFactoryClass() != null &&
+					mailProps.getSmtpAuth() != null)
+				{
+					mailResource.setProperty("mail.smtp.from", mailProps.getSmtpFrom());
+					mailResource.setProperty("mail.smtp.password", mailProps.getSmtpPassword());
+					mailResource.setProperty("mail.smtp.starttls.enable", mailProps.getSmtpStartTlsEnable());
+					mailResource.setProperty("mail.smtp.socketFactory.class", mailProps.getSmtpSocketFactoryClass());
+					mailResource.setProperty("mail.smtp.auth", mailProps.getSmtpAuth());
+				}
+
 				return mailResource;
 			}
 		};
 	}
 
-	/**
-	 * Extracts a zip file specified by the zipFilePath to a directory specified by
-	 * destDirectory (will be created if does not exists)
-	 * @param zipInputStream
-	 * @param destDirectory
-	 * @throws IOException
-	 */
-	public static void extractZip(InputStream zipInputStream, String destDirectory) throws IOException
+	private static void extractExecutableJar(String destDirectory, String jarFilePath)
+	{
+		try
+		{
+			// check for multiple jars and blow up if multiple are present
+			try (JarFile jar = new JarFile(jarFilePath))
+			{
+				boolean foundDistributionZip = false;
+				var entries = jar.entries();
+				while (entries.hasMoreElements())
+				{
+					var entry = entries.nextElement();
+					var entryName = entry.getName();
+
+					if ("labkey/distribution.zip".equals(entryName))
+					{
+						foundDistributionZip = true;
+						try (var distInputStream = jar.getInputStream(entry))
+						{
+							LabKeyServer.extractZip(distInputStream, destDirectory);
+						}
+					}
+				}
+
+				if (!foundDistributionZip)
+				{
+					throw new ConfigException("Unable to find distribution zip required to run LabKey server.");
+				}
+			}
+		}
+		catch (IOException | ConfigException e)
+		{
+			throw new RuntimeException(e);
+		}
+	}
+
+	private static String getExecutableJar(String currentPath) throws ConfigException
+	{
+		File currentDir = new File(currentPath);
+		var jarCount = 0;
+		String jarFileName = "";
+
+		for (File file: currentDir.listFiles())
+		{
+			if (file.getName().endsWith(".jar"))
+			{
+				jarFileName = file.getName();
+				jarCount++;
+			}
+		}
+
+		if (jarCount == 0)
+		{
+			throw new ConfigException("Executable jar not found.");
+		}
+
+		// only 1 executable jar should be there
+		if (jarCount == 1)
+		{
+			return jarFileName;
+		}
+
+		throw new ConfigException("Multiple executable jars found. Must provide only one executable jar");
```

Comment:
Why not build a Set<String> so we can give an even better error message?

And since we're not checking if the JAR is actually executable, I'd just say "jar"